nexus/src/app/sagas/region_snapshot_replacement_start.rs

```diff
@@ -0,0 +1,1128 @@
+            params.request.old_region_id,
+            params.request.old_snapshot_id,
+        )));
+    };
+
+    let old_snapshot_address: SocketAddrV6 =
+        match region_snapshot.snapshot_addr.parse() {
+            Ok(addr) => addr,
+
+            Err(e) => {
+                return Err(ActionError::action_failed(format!(
+                    "parsing {} as SocketAddrV6 failed: {e}",
+                    region_snapshot.snapshot_addr,
+                )));
+            }
+        };
+
+    let (new_dataset, ensured_region) = sagactx.lookup::<(
+        db::model::Dataset,
+        crucible_agent_client::types::Region,
+    )>(
+        "ensured_dataset_and_region",
+    )?;
+
+    let Some(new_dataset_address) = new_dataset.address() else {
+        return Err(ActionError::action_failed(format!(
+            "dataset {} does not have an address!",
+            new_dataset.id(),
+        )));
+    };
+
+    let new_region_address = SocketAddrV6::new(
+        *new_dataset_address.ip(),
+        ensured_region.port_number,
+        0,
+        0,
+    );
+
+    let old_volume_id = sagactx.lookup::<Uuid>("old_snapshot_volume_id")?;
+
+    // Return the replacement parameters for the forward action case - the undo
+    // will swap the existing and replacement target
+    Ok(ReplaceParams {
+        old_volume_id,
+        old_snapshot_address,
+        new_region_address,
+        new_volume_id,
+    })
+}
+
+async fn rsrss_replace_snapshot_in_volume(
+    sagactx: NexusActionContext,
+) -> Result<(), ActionError> {
+    let log = sagactx.user_data().log();
+    let osagactx = sagactx.user_data();
+
+    let replacement_params = get_replace_params(&sagactx).await?;
+
+    info!(
+        log,
+        "replacing {} with {} in volume {}",
+        replacement_params.old_snapshot_address,
+        replacement_params.new_region_address,
+        replacement_params.old_volume_id,
+    );
+
+    // `volume_replace_snapshot` will swap the old snapshot for the new region.
+    // No repair or reconcilation needs to occur after this.
+    osagactx
+        .datastore()
+        .volume_replace_snapshot(
+            VolumeWithTarget(replacement_params.old_volume_id),
+            ExistingTarget(replacement_params.old_snapshot_address),
+            ReplacementTarget(replacement_params.new_region_address),
+            VolumeToDelete(replacement_params.new_volume_id),
+        )
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    Ok(())
+}
+
+async fn rsrss_replace_snapshot_in_volume_undo(
+    sagactx: NexusActionContext,
+) -> Result<(), anyhow::Error> {
+    // Undo the forward action's volume_replace_snapshot call by swapping the
+    // existing target and replacement target parameters.
+
+    let log = sagactx.user_data().log();
+    let osagactx = sagactx.user_data();
+
+    let replacement_params = get_replace_params(&sagactx).await?;
+
+    info!(
+        log,
+        "undo: replacing {} with {} in volume {}",
```

Comment:
At this point is "old" from the broken region we were fixing?  Or have we swapped
old/new here?  I'm not sure if this log is printing what I think it should be, which is
opposite from the `srss_replace_snapshot_in_volume()`.

Or, wait, by saying `undo` at the top, you mean we are undo-ing what the previous
step did, which is putting back things like they were.  I guess I find this a little
confusing, but if all the `undo` messages are like this, I'll just live with it.

Reply:
The old and new is _not_ swapped in this log message, where it is in the `volume_replace_snapshot` arguments. You're right, adding `undo: ` is meant to signify that it's undoing the forward action. The intention is that someone can search for the text "replacing UUID with UUID in volume UUID" and get two results.

I put a small comment in 1b975251b explaining this intention.